logbook-servlet/src/main/java/org/zalando/logbook/servlet/RemoteRequest.java

```diff
@@ -89,10 +95,33 @@ public Charset getCharset() {
 
     @Override
     public HttpRequest withBody() throws IOException {
+        if (isFormRequest()) {
+            switch (formRequestMode) {
+                case PARAMETER:
+                    this.body = reconstructBodyFromParameters();
+                    return this;
+                case OFF:
+                    this.body = EMPTY_BODY;
+                    return this;
+            }
+        }
+
         body = ByteStreams.toByteArray(super.getInputStream());
         return this;
     }
 
+    private boolean isFormRequest() {
+        return "application/x-www-form-urlencoded".equals(getContentType());
+    }
+
+    private byte[] reconstructBodyFromParameters() {
+        return getParameterMap().entrySet().stream()
+                .flatMap(entry -> Arrays.stream(entry.getValue())
```

Comment:
Key and entry should be escaped (imagine `name=Johnson & Johnson`)

Reply:
We can't be 100% certain how the parameters were passed to us in the first place. They could come from the query string as well as the body. In addition, we don't know whether they were encoded. It's a judgement call now, since in the end we want to log them and present them to humans (unencoded, since that would be easier) but on the other hand we want to preserve the format on the wire (encoded, most likely).

Reply:
If we log real body, how would it look like? Should reconstructed be as close as possible in its form to real one?

Reply:
> If we log real body, how would it look like?

Untouched, i.e. encoded.

> Should reconstructed be as close as possible in its form to real one?

Yes.